arch/rv32i/src/lib.rs

```diff
@@ -337,10 +337,10 @@ pub extern "C" fn _start_trap() {
             sw   t0, 1*4(s0)  // Save the app sp to the stored state struct
             csrr t0, 0x341    // CSR=0x341=mepc
             sw   t0, 31*4(s0) // Save the PC to the stored state struct
-            csrr t0, 0x342    // CSR=0x342=mcause
-            sw   t0, 32*4(s0) // Save mcause to the stored state struct
             csrr t0, 0x343    // CSR=0x343=mtval
             sw   t0, 33*4(s0) // Save mtval to the stored state struct
+            csrr t0, 0x342    // CSR=0x342=mcause
```

Comment:
Good catch @vsukhoml! I think it might be worth adding an additional comment here which indicates why exactly we're following this order. I'd argue in such a section of mechanical CSR saves this is really easy to miss and likely why the issue was introduced in the first place.

```suggestion

            // Save mcause last, as we depend on it being loaded in t0 below
            csrr t0, 0x342    // CSR=0x342=mcause
```